// ts-add-js-extension/build/cjs/read-write.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeMany = exports.findMany = void 0;
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var typescript_1 = __importDefault(require("typescript"));
var const_1 = require("./const");
var log_1 = __importDefault(require("./log"));
var traverse_and_update_1 = __importDefault(require("./traverse-and-update"));
var getAllJSAndDTSFiles = function (directory) {
    return fs_1.default.readdirSync(directory).flatMap(function (file) {
        var filePath = path_1.default.posix.join(directory.split(path_1.default.sep).join(const_1.separator), file);
        if (fs_1.default.statSync(filePath).isDirectory()) {
            return getAllJSAndDTSFiles(filePath);
        }
        return !(0, const_1.matchEither)(filePath) ? [] : [filePath];
    });
};
var readCode = function (file) {
    return new Promise(function (resolve, reject) {
        var code = [];
        fs_1.default.createReadStream(file)
            .on('data', function (data) {
            code.push(data.toString());
        })
            .on('end', function () {
            resolve(code.join(''));
        })
            .on('error', reject);
    });
};
var getAllJSAndDTSMetainfo = function (files) {
    return files.map(function (file) { return __awaiter(void 0, void 0, void 0, function () {
        var code;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, readCode(file)];
                case 1:
                    code = _a.sent();
                    return [2 /*return*/, {
                            files: files,
                            code: code,
                            sourceFile: typescript_1.default.createSourceFile(file, code, typescript_1.default.ScriptTarget.ESNext),
                        }];
            }
        });
    }); });
};
var findMany = function (props) { return __awaiter(void 0, void 0, void 0, function () {
    var files;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                files = props.include.concat(props.dir).flatMap(getAllJSAndDTSFiles);
                return [4 /*yield*/, Promise.all(getAllJSAndDTSMetainfo(files))];
            case 1: return [2 /*return*/, _a.sent()];
        }
    });
}); };
exports.findMany = findMany;
var writeMany = function (props) { return __awaiter(void 0, void 0, void 0, function () {
    var transformed, repeat, log, errors, error_1;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                transformed = props.foundMany.flatMap(traverse_and_update_1.default);
                repeat = transformed.reduce(function (longestFileName, _a) {
                    var file = _a.file;
                    return Math.max(longestFileName, file.length);
                }, 0);
                log = !(transformed.length && props.showProgress)
                    ? undefined
                    : log_1.default.fromNumberOfFiles(transformed.length);
                _a.label = 1;
            case 1:
                _a.trys.push([1, 3, , 4]);
                return [4 /*yield*/, Promise.all(transformed.map(function (_a) {
                        var code = _a.code, file = _a.file;
                        return new Promise(function (resolve) {
                            return fs_1.default.writeFile(file, code, function (error) {
                                log === null || log === void 0 ? void 0 : log.increment({
                                    repeat: repeat,
                                    file: file,
                                    succeed: !error,
                                });
                                resolve(!error
                                    ? undefined
                                    : {
                                        file: file,
                                        error: error,
                                    });
                            });
                        });
                    }))];
            case 2:
                errors = (_a.sent()).flatMap(function (element) {
                    return !element ? [] : [element];
                });
                log === null || log === void 0 ? void 0 : log.end({ errors: errors });
                return [2 /*return*/, {
                        type: 'done',
                    }];
            case 3:
                error_1 = _a.sent();
                return [2 /*return*/, {
                        type: 'error',
                        error: error_1,
                    }];
            case 4: return [2 /*return*/];
        }
    });
}); };
exports.writeMany = writeMany;
